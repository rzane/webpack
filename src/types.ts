import { Configuration, Entry } from "webpack";

export type { Configuration, Entry };
export type Mode = "development" | "production";
export type Hook = (context: Configuration) => Configuration;

export interface ModeOptions {
  default?: Hook;
  development?: Hook;
  production?: Hook;
}

export interface OutputOptions {
  path: string;
  publicPath?: string;
}

export interface FilesOptions {
  test: RegExp;
}
