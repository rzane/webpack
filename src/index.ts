export * from "./hooks";
export * from "./pipeline";
export * from "./types";
