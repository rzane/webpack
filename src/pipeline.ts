import { Hook, Mode, Configuration } from "./types";

/**
 * Create a webpack configuration factory from a set of hooks
 */
export const pipeline = (hooks: Hook[]) => (env: Mode): Configuration => {
  /**
   * Set the `NODE_ENV` for downstream tools to consume.
   */
  process.env.NODE_ENV = env;

  /**
   * Build our initial configuration object
   */
  const config: Configuration = {
    mode: env,
  };

  /**
   * Reduce over each hook, updating the configuration
   */
  return hooks.reduce((acc, hook) => hook(acc), config);
};
