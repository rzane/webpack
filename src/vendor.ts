import * as path from "path";

const VENDOR_NAME = "vendor";
const MODULES = "node_modules";

/**
 * @private
 */
export const VENDOR_CONFIG = {
  name: VENDOR_NAME,
  enforce: true,
  chunks: "all" as const,
  test: /[\\/]node_modules[\\/]/,
};

/**
 * @private
 */
export function getVendorName(filename: string) {
  const segments = path.normalize(filename).split(path.sep);
  const index = segments.indexOf(MODULES);

  let name = segments[index + 1];
  if (name.startsWith("@")) {
    name = `${name.slice(1)}__${segments[index + 2]}`;
  }

  return path.join(VENDOR_NAME, name);
}
