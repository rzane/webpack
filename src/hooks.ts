import { HotModuleReplacementPlugin } from "webpack";
import { merge as _merge } from "webpack-merge";
import CompressionPlugin from "compression-webpack-plugin";
import HtmlWebpackPlugin from "html-webpack-plugin";
import MiniCssExtractPlugin from "mini-css-extract-plugin";
import CssMinimizerPlugin from "css-minimizer-webpack-plugin";
import FaviconsWebpackPlugin from "favicons-webpack-plugin";
import ReactRefreshWebpackPlugin from "@pmmmwh/react-refresh-webpack-plugin";
import { getVendorName, VENDOR_CONFIG } from "./vendor";
import {
  Configuration,
  Entry,
  FaviconOptions,
  FilesOptions,
  HTMLOptions,
  Hook,
  Mode,
  ModeOptions,
  OutputOptions,
  Plugin,
  Rule,
  PostCSSOptions,
} from "./types";

const assert = (test: any, message: string) => {
  if (!test) {
    throw new Error(message);
  }
};

/**
 * Add custom configuration to your Webpack config.
 * @public
 * @example
 * merge({
 *   module: {
 *     rules: [{ test: /\.txt$/, use: 'raw-loader' }],
 *   },
 * })
 */
export const merge = (config: Configuration): Hook => (previousConfig) => {
  return _merge(previousConfig, config);
};

/**
 * Provide environment-specific configuration.
 * @public
 * @example
 * mode({
 *   development: merge({ devtool: "cheap-module-source-map" })
 * })
 */
export const mode = (options: ModeOptions): Hook => (config) => {
  return [options.default, options[config.mode as Mode]]
    .filter((hook): hook is Hook => typeof hook !== "undefined")
    .reduce((acc, fn) => fn(acc), config);
};

/**
 * Configure the webpack entrypoint or entrypoints.
 *
 * See the [Webpack documentation](https://webpack.js.org/concepts/entry-points/)
 * for more information.
 *
 * @public
 * @example <caption>A single entrypoint</caption>
 * entry("src/index.ts")
 * @example <caption>Multiple entrypoints</caption>
 * entry(["src/index.ts", "src/admin.ts"])
 * @example <caption>Multiple named entrypoints</caption>
 * entry({
 *   main: "src/index.ts",
 *   admin: "src/admin.ts"
 * })
 */
export const entry = (entry: string | string[] | Entry): Hook => {
  return merge({ entry });
};

/**
 * Set the output directory
 *
 * See the [Webpack documentation](https://webpack.js.org/concepts/output/)
 * for more information.
 *
 * @public
 * @example
 * output({
 *   path: path.join(__dirname, "dist"),
 *   publicPath: "/",
 * })
 */
export const output = (options: OutputOptions): Hook => {
  assert(options.path, "`output` expects a `path` property");

  return mode({
    development: merge({
      output: {
        publicPath: options.publicPath,
        filename: "assets/js/[name].js",
        chunkFilename: "assets/js/[name].chunk.js",
      },
    }),
    production: merge({
      output: {
        path: options.path,
        publicPath: options.publicPath,
        filename: "assets/js/[name].[contenthash:8].js",
        chunkFilename: "assets/js/[name].[contenthash:8].chunk.js",
      },
    }),
  });
};

/**
 * Add a new rule
 * @public
 * @example
 * rule({
 *   test: /\.txt$/,
 *   use: "raw-loader"
 * })
 */
export const rule = (rule: Rule): Hook => {
  return merge({ module: { rules: [rule] } });
};

/**
 * Add a new plugin
 * @public
 * @example
 * plugin(new BannerPlugin({ banner: "Hello!" }))
 */
export const plugin = (plugin: Plugin): Hook => {
  return merge({ plugins: [plugin] });
};

/**
 * Build JavaScript and TypeScript with Babel
 * @public
 * @example
 * babel()
 */
export const babel = (): Hook => {
  return merge({
    module: {
      rules: [
        {
          test: /\.(js|mjs|jsx|ts|tsx)$/,
          exclude: /node_modules/,
          use: [
            {
              loader: require.resolve("babel-loader"),
              options: { cacheDirectory: true },
            },
          ],
        },
      ],
    },
    resolve: {
      extensions: [".tsx", ".ts", ".mjs", ".js"],
    },
  });
};

/**
 * Compile CSS with PostCSS
 * @public
 * @example
 * postcss()
 */
export const postcss = (opts: PostCSSOptions = {}): Hook => {
  const test = /\.(css|pcss|postcss)$/;

  const css = {
    loader: require.resolve("css-loader"),
    options: { ...opts, importLoaders: 1 },
  };

  const postcss = {
    loader: require.resolve("postcss-loader"),
    options: { ...opts, postcssOptions: { hideNothingWarning: true } },
  };

  return mode({
    development: merge({
      module: {
        rules: [
          {
            test,
            use: [require.resolve("style-loader"), css, postcss],
          },
        ],
      },
    }),
    production: merge({
      module: {
        rules: [
          {
            test,
            use: [MiniCssExtractPlugin.loader, css, postcss],
          },
        ],
      },
      plugins: [
        new MiniCssExtractPlugin({
          filename: "assets/css/[name].[contenthash:8].css",
          chunkFilename: "assets/css/[name].[contenthash:8].chunk.css",
        }),
      ],
    }),
  });
};

/**
 * Allows SVG files to be imported as React components or as URLs.
 * @public
 * @example <caption>Enabling the hook</caption>
 * svg()
 * @example <caption>Importing a React component</caption>
 * import { ReactComponent as Logo } from "./logo.svg";
 *
 * <Logo />
 * @example <caption>Importing a URL to an SVG</caption>
 * import url from "./logo.svg";
 *
 * <img src={url} />
 */
export const svg = (): Hook => {
  return rule({
    test: /\.svg$/,
    use: [
      require.resolve("@svgr/webpack"),
      {
        loader: require.resolve("file-loader"),
        options: { name: "assets/media/[name].[hash:8].[ext]" },
      },
    ],
  });
};

/**
 * Allows files to be imported.
 * @public
 * @example <caption>Enabling the hook</caption>
 * files({ test: /\.(png|jpg)$/ })
 * @example <caption>Importing a file</caption>
 * import url from "./logo.png";
 *
 * <img src={url} />
 */
export const files = (options: FilesOptions): Hook => {
  assert(options.test, "`files` expects a `test` property");

  return rule({
    test: options.test,
    use: [
      {
        loader: require.resolve("file-loader"),
        options: { name: "assets/media/[name].[hash:8].[ext]" },
      },
    ],
  });
};

/**
 * Produce an HTML file.
 * @public
 * @example
 * html()
 */
export const html = (options: HTMLOptions = {}): Hook => {
  return plugin(
    new HtmlWebpackPlugin({
      scriptLoading: "defer",
      ...options,
    })
  );
};

/**
 * Generate favicons and a web app manifest.
 * @public
 * @example
 * favicons({
 *   name: "Example",
 *   logo: "assets/logo.png",
 *   background: "#eee",
 *   foreground: "#000",
 * })
 */
export const favicons = (options: FaviconOptions): Hook => {
  assert(options.name, "`favicons` expected a `name` property");
  assert(options.logo, "`favicons` expected a `logo` property");

  const faviconPlugin = new FaviconsWebpackPlugin({
    logo: options.logo,
    prefix: "assets/icons/",
    cache: true,
    mode: "webapp",
    devMode: "light",
    favicons: {
      appName: options.name,
      appShortName: options.shortName,
      appDescription: options.description,
      developerURL: null,
      background: options.background || "#fff",
      theme_color: options.themeColor || "#fff",
      icons: { coast: false, yandex: false },
    },
  });

  return plugin(faviconPlugin);
};

/**
 * After using this hook, you'll have a single file that contains all of the
 * node modules used in your application. For example:
 *
 *    assets/js/main.js
 *    assets/js/vendor.js
 *
 * This increases the likelihood of a cache hit on subsequent deploys. If none
 * of your dependencies change, your users can use a cached version of the
 * vendors file.
 *
 * This hook is mutually exclusive with the [vendorEachModule](#vendoreachmodule) hook.
 *
 * @public
 * @example
 * vendor()
 */
export const vendor = (): Hook => {
  return mode({
    production: merge({
      optimization: {
        runtimeChunk: "single",
        splitChunks: {
          cacheGroups: {
            vendor: VENDOR_CONFIG,
          },
        },
      },
    }),
  });
};

/**
 * After using this hook, you'll have a file for each top-level node module in
 * your application. For example:
 *
 *     assets/js/main.js
 *     assets/js/vendor/react.js
 *     assets/js/vendor/react-router.js
 *     assets/js/vendor/react-router-dom.js
 *
 * This maximizes the likelihood of a cache hit on subsequent deploys, but it
 * also results in a lot of HTTP requests. If your server doesn't support
 * HTTP/2, you are probably better off using the [vendor](#vendor) hook.
 *
 * This hook is mutually exclusive with the [vendor](#vendor) hook.
 *
 * @public
 * @example
 * vendorEachModule()
 */
export const vendorEachModule = (): Hook => {
  return mode({
    production: merge({
      optimization: {
        runtimeChunk: "single",
        splitChunks: {
          cacheGroups: {
            vendor: {
              ...VENDOR_CONFIG,
              name: (mod: any) => getVendorName(mod.context),
            },
          },
        },
      },
    }),
  });
};

/**
 * Minify resulting JavaScript and CSS.
 * @public
 * @example
 * minify()
 */
export const minify = (): Hook => {
  return mode({
    production: merge({
      optimization: {
        minimize: true,
        minimizer: ["...", new CssMinimizerPlugin()],
      },
    }),
  });
};

/**
 * Enable GZIP compression.
 * @public
 * @example
 * gzip()
 */
export const gzip = (): Hook => {
  return mode({ production: merge({ plugins: [new CompressionPlugin()] }) });
};

/**
 * Enable fast refresh for React. This depends on using the `hot` option
 * for `@stackup/webpack/babel-preset`.
 *
 * @public
 * @example
 * fastRefresh()
 */
export const refresh = (): Hook => {
  return mode({
    development: merge({
      plugins: [
        new HotModuleReplacementPlugin(),
        new ReactRefreshWebpackPlugin(),
      ],
    }),
  });
};
