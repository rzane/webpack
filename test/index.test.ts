import * as webpack from "../src";
import MiniCssExtractPlugin from "mini-css-extract-plugin";
import HtmlWebpackPlugin from "html-webpack-plugin";
import CompressionPlugin from "compression-webpack-plugin";
import FaviconsWebpackPlugin from "favicons-webpack-plugin";

const rule: webpack.Rule = {
  test: /foo/,
  use: [],
};

const plugin: webpack.Plugin = {
  apply: () => undefined,
};

const build = webpack.pipeline([
  webpack.entry({ app: "/tmp/a" }),
  webpack.output({ path: "dist", publicPath: "/" }),
  webpack.babel(),
  webpack.postcss(),
  webpack.svg(),
  webpack.html(),
  webpack.vendor(),
  webpack.minify(),
  webpack.gzip(),
  webpack.rule(rule),
  webpack.files({ test: /\.mp4/ }),
  webpack.plugin(plugin),
]);

const getLoaders = (config: any) => {
  const loaders: string[] = [];
  for (const rule of config.module.rules) {
    for (const use of rule.use) {
      if (typeof use === "string") {
        loaders.push(use);
      } else {
        loaders.push(use.loader);
      }
    }
  }
  return loaders;
};

test("entry", () => {
  const config = build({});
  expect(config.entry).toEqual({ app: "/tmp/a" });
});

describe("output", () => {
  test("development", () => {
    const config = build({});
    expect(config.output!.path).toBeUndefined();
    expect(config.output!.publicPath).toEqual("/");
    expect(config.output!.filename).not.toContain("contenthash");
  });

  test("production", () => {
    const config = build({ production: true });
    expect(config.output!.path).toEqual("dist");
    expect(config.output!.publicPath).toEqual("/");
    expect(config.output!.filename).toContain("contenthash");
  });
});

test("babel", () => {
  const config = build({});
  const loaders = getLoaders(config);
  expect(loaders).toContainEqual(expect.stringContaining("babel-loader"));
});

describe("postcss", () => {
  test("development", () => {
    const config = build({});
    const loaders = getLoaders(config);
    expect(loaders).toContainEqual(expect.stringContaining("style-loader"));
    expect(loaders).toContainEqual(expect.stringContaining("postcss-loader"));
  });

  test("production", () => {
    const config = build({ production: true });
    const loaders = getLoaders(config);
    expect(loaders).not.toContainEqual(expect.stringContaining("style-loader"));
    expect(loaders).toContainEqual(expect.stringContaining("postcss-loader"));
    expect(config.plugins).toContainEqual(expect.any(MiniCssExtractPlugin));
  });
});

test("svg", () => {
  const config = build({});
  const loaders = getLoaders(config);
  expect(loaders).toContainEqual(expect.stringContaining("@svgr/webpack"));
  expect(loaders).toContainEqual(expect.stringContaining("file-loader"));
});

test("files", () => {
  const config = build({});
  const loaders = getLoaders(config);
  expect(loaders).toContainEqual(expect.stringContaining("file-loader"));
});

test("html", () => {
  const config = build({});
  expect(config.plugins).toContainEqual(expect.any(HtmlWebpackPlugin));
});

describe("vendor", () => {
  test("development", () => {
    const config = build({});
    expect(config.optimization).toBeUndefined();
  });

  test("production", () => {
    const config = build({ production: true });
    expect(config.optimization!.splitChunks).toBeDefined();
  });
});

describe("minify", () => {
  test("development", () => {
    const config = build({});
    expect(config.optimization).toBeUndefined();
  });

  test("production", () => {
    const config = build({ production: true });
    expect(config.optimization!.minimize).toBeDefined();
  });
});

describe("gzip", () => {
  test("development", () => {
    const config = build({});
    expect(config.plugins).not.toContainEqual(expect.any(CompressionPlugin));
  });

  test("production", () => {
    const config = build({ production: true });
    expect(config.plugins).toContainEqual(expect.any(CompressionPlugin));
  });
});

test("favicons", () => {
  const config = build({});
  expect(config.plugins).toContainEqual(expect.any(FaviconsWebpackPlugin));
});

test("rule", () => {
  const config = build({});
  expect(config.module!.rules).toContainEqual(rule);
});

test("plugin", () => {
  const config = build({});
  expect(config.plugins).toContain(plugin);
});
