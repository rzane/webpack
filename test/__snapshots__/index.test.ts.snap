// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`builds a development config 1`] = `
  Object {
    "entry": Object {
      "app": "/tmp/a",
    },
    "mode": "development",
    "module": Object {
      "rules": Array [
        Object {
          "exclude": /node_modules/,
          "test": /\\\\\\.\\(js\\|mjs\\|jsx\\|ts\\|tsx\\)\\$/,
          "use": Array [
            Object {
              "loader": "/node_modules/babel-loader/lib/index.js",
              "options": Object {
                "cacheDirectory": true,
              },
            },
          ],
        },
        Object {
          "test": /\\\\\\.css\\$/,
          "use": Array [
            "/node_modules/style-loader/dist/cjs.js",
            Object {
              "loader": "/node_modules/css-loader/dist/cjs.js",
              "options": Object {
                "importLoaders": 1,
              },
            },
            "/node_modules/postcss-loader/src/index.js",
          ],
        },
        Object {
          "test": /\\\\\\.svg\\$/,
          "use": Array [
            "/node_modules/@svgr/webpack/lib/index.js",
            Object {
              "loader": "/node_modules/file-loader/dist/cjs.js",
              "options": Object {
                "name": "assets/media/[name].[hash:8].[ext]",
              },
            },
          ],
        },
        Object {
          "test": /\\\\\\.mp4/,
          "use": Array [
            Object {
              "loader": "/node_modules/file-loader/dist/cjs.js",
              "options": Object {
                "name": "assets/media/[name].[hash:8].[ext]",
              },
            },
          ],
        },
      ],
    },
    "output": Object {
      "chunkFilename": "assets/js/[name].chunk.js",
      "filename": "assets/js/[name].js",
      "publicPath": "/",
    },
    "plugins": Array [
      HtmlWebpackPlugin {
        "assetJson": undefined,
        "childCompilerHash": undefined,
        "hash": undefined,
        "options": Object {
          "base": false,
          "cache": true,
          "chunks": "all",
          "chunksSortMode": "auto",
          "compile": true,
          "excludeChunks": Array [],
          "favicon": false,
          "filename": "index.html",
          "hash": false,
          "inject": "body",
          "meta": Object {
            "viewport": "width=device-width, initial-scale=1",
          },
          "minify": "auto",
          "scriptLoading": "blocking",
          "showErrors": true,
          "template": "auto",
          "templateContent": false,
          "templateParameters": [Function],
          "title": "Webpack App",
          "xhtml": false,
        },
        "version": 4,
      },
      FaviconsWebpackPlugin {
        "options": Object {
          "cache": true,
          "favicons": Object {
            "appDescription": undefined,
            "appName": "Example",
            "appShortName": undefined,
            "background": "#fff",
            "developerURL": null,
            "icons": Object {
              "coast": false,
              "yandex": false,
            },
            "theme_color": "#fff",
          },
          "inject": true,
          "logo": "src/assets/logo.png",
          "prefix": "assets/icons/",
        },
      },
    ],
    "resolve": Object {
      "extensions": Array [
        ".tsx",
        ".ts",
        ".mjs",
        ".js",
      ],
    },
  }
`;

exports[`builds a production config 1`] = `
  Object {
    "entry": Object {
      "app": "/tmp/a",
    },
    "mode": "production",
    "module": Object {
      "rules": Array [
        Object {
          "exclude": /node_modules/,
          "test": /\\\\\\.\\(js\\|mjs\\|jsx\\|ts\\|tsx\\)\\$/,
          "use": Array [
            Object {
              "loader": "/node_modules/babel-loader/lib/index.js",
              "options": Object {
                "cacheDirectory": true,
              },
            },
          ],
        },
        Object {
          "test": /\\\\\\.css\\$/,
          "use": Array [
            "/node_modules/mini-css-extract-plugin/dist/loader.js",
            Object {
              "loader": "/node_modules/css-loader/dist/cjs.js",
              "options": Object {
                "importLoaders": 1,
              },
            },
            "/node_modules/postcss-loader/src/index.js",
          ],
        },
        Object {
          "test": /\\\\\\.svg\\$/,
          "use": Array [
            "/node_modules/@svgr/webpack/lib/index.js",
            Object {
              "loader": "/node_modules/file-loader/dist/cjs.js",
              "options": Object {
                "name": "assets/media/[name].[hash:8].[ext]",
              },
            },
          ],
        },
        Object {
          "test": /\\\\\\.mp4/,
          "use": Array [
            Object {
              "loader": "/node_modules/file-loader/dist/cjs.js",
              "options": Object {
                "name": "assets/media/[name].[hash:8].[ext]",
              },
            },
          ],
        },
      ],
    },
    "optimization": Object {
      "minimize": true,
      "minimizer": Array [
        TerserPlugin {
          "options": Object {
            "cache": true,
            "cacheKeys": [Function],
            "exclude": undefined,
            "extractComments": false,
            "include": undefined,
            "minify": undefined,
            "parallel": true,
            "sourceMap": undefined,
            "terserOptions": Object {},
            "test": /\\\\\\.m\\?js\\(\\\\\\?\\.\\*\\)\\?\\$/i,
            "warningsFilter": [Function],
          },
        },
        OptimizeCssAssetsWebpackPlugin {
          "deleteAssetsMap": Object {},
          "options": Object {
            "assetNameRegExp": /\\\\\\.css\\(\\\\\\?\\.\\*\\)\\?\\$/i,
            "assetProcessors": Array [
              Object {
                "phase": "compilation.optimize-chunk-assets",
                "processor": [Function],
                "regExp": /\\\\\\.css\\(\\\\\\?\\.\\*\\)\\?\\$/i,
              },
            ],
            "canPrint": undefined,
            "cssProcessor": [Function],
            "cssProcessorOptions": Object {},
            "cssProcessorPluginOptions": Object {},
          },
          "phaseAssetProcessors": Object {
            "compilation.optimize-assets": Array [],
            "compilation.optimize-chunk-assets": Array [
              Object {
                "phase": "compilation.optimize-chunk-assets",
                "processor": [Function],
                "regExp": /\\\\\\.css\\(\\\\\\?\\.\\*\\)\\?\\$/i,
              },
            ],
            "emit": Array [],
          },
          "pluginDescriptor": Object {
            "name": "OptimizeCssAssetsWebpackPlugin",
          },
        },
      ],
      "runtimeChunk": "single",
      "splitChunks": Object {
        "cacheGroups": Object {
          "vendor": Object {
            "chunks": "all",
            "enforce": true,
            "name": "vendors",
            "test": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]/,
          },
        },
      },
    },
    "output": Object {
      "chunkFilename": "assets/js/[name].[contenthash:8].chunk.js",
      "filename": "assets/js/[name].[contenthash:8].js",
      "path": "/tmp/b",
      "publicPath": "/",
    },
    "plugins": Array [
      MiniCssExtractPlugin {
        "options": Object {
          "chunkFilename": "assets/css/[name].[contenthash:8].chunk.css",
          "filename": "assets/css/[name].[contenthash:8].css",
          "ignoreOrder": false,
          "moduleFilename": [Function],
        },
      },
      HtmlWebpackPlugin {
        "assetJson": undefined,
        "childCompilerHash": undefined,
        "hash": undefined,
        "options": Object {
          "base": false,
          "cache": true,
          "chunks": "all",
          "chunksSortMode": "auto",
          "compile": true,
          "excludeChunks": Array [],
          "favicon": false,
          "filename": "index.html",
          "hash": false,
          "inject": "body",
          "meta": Object {
            "viewport": "width=device-width, initial-scale=1",
          },
          "minify": "auto",
          "scriptLoading": "blocking",
          "showErrors": true,
          "template": "auto",
          "templateContent": false,
          "templateParameters": [Function],
          "title": "Webpack App",
          "xhtml": false,
        },
        "version": 4,
      },
      CompressionPlugin {
        "emittedAssets": Set {},
        "options": Object {
          "algorithm": [Function],
          "cache": true,
          "compressionOptions": Object {
            "level": 9,
          },
          "deleteOriginalAssets": false,
          "exclude": undefined,
          "filename": "[path].gz[query]",
          "include": undefined,
          "minRatio": 0.8,
          "test": undefined,
          "threshold": 0,
        },
      },
      FaviconsWebpackPlugin {
        "options": Object {
          "cache": true,
          "favicons": Object {
            "appDescription": undefined,
            "appName": "Example",
            "appShortName": undefined,
            "background": "#fff",
            "developerURL": null,
            "icons": Object {
              "coast": false,
              "yandex": false,
            },
            "theme_color": "#fff",
          },
          "inject": true,
          "logo": "src/assets/logo.png",
          "prefix": "assets/icons/",
        },
      },
    ],
    "resolve": Object {
      "extensions": Array [
        ".tsx",
        ".ts",
        ".mjs",
        ".js",
      ],
    },
  }
`;
